Support `@Scheduled` on Reactive methods and Kotlin suspending functions

simonbasle opened the pull request:
This commit adds support for `@Scheduled` annotation on reactive methods
and on Kotlin suspending functions (provided a bridge to Reactive Streams
is available at runtime).

Reactive methods are methods that return an instance of Publisher or an
instance of a class that the ReactiveAdapterRegistry can adapt to a
Publisher with deferred support.

Kotlin suspending functions are converted to Publisher as well, using
`CoroutineUtils` and the support of the `kotlinx.coroutines.reactor`
bridge.

The bean method that produces the `Publisher` is only called once, but
there can be multiple subscriptions to that instance as configured by
the annotation. The usual task-based infrastructure for synchronous code
is reused for the purpose of scheduling the subscriptions.

One special case is when a `fixedDelay` is used, as the non-blocking
nature of subscribing to a Publisher makes it harder to adhere to these
semantics. As a result, for that particular case only, the subscription
is done in a blocking fashion inside the scheduled task.

Publisher onNext events are ignored. Active subscriptions are tracked
by the processor so that long-running publishers and infinite publishers
are also supported, allowing for cancellation if the associated bean is
destroyed or if the context is stopped.

See gh-23533
Closes gh-29924

simonbasle committed:
Support `@Scheduled` fixedDelay/fixedRate on Publisher-returning methods

This commit adds support for `@Scheduled` annotation on reactive methods
in fixed delay or fixed rate mode. Cron mode is not supported.

Reactive methods are methods that return a Publisher or a subclass of
Publisher. This is only considered if Reactor (and Reactive Streams)
is present at runtime.

This is implemented using Reactor operators, as a `Flux<Void>` that
repeatedly flatmaps to the `Publisher` in the background, re-subscribing
to it according to the `@Scheduled` configuration. The method which
creates the `Publisher` is only called once.

If the `Publisher` errors, the exception is logged at warn level and
otherwise ignored. As a result the underlying `Flux` will continue
re-subscribing to the `Publisher`.

Note that if Reactor is not present, the standard processing applies
and the method itself will repeatedly be scheduled for execution,
ignoring the returned `Publisher` and not even subscribing to it. This
effectively becomes a no-op operation unless the method has other side
effects.

Closes gh-23533

simonbasle committed:
Add tests + fix initialDelay.ZERO case

simonbasle committed:
fix checkpoint test and checkstyle polish

sdeleuze commented:
We should probably review #28515 at the same time and have a common approach.

sdeleuze commented:
As discussed, let's implement the Coroutines support as part of this PR via an invocation of `.CoroutinesUtils#invokeSuspendingFunction(Method, Object, Object...)` when `KotlinDetector#isSuspendingFunction` detect a method as suspending, like we do for the Web support + related tests in a `KotlinXxx.kt` reusing the same logic from Java tests with suspending function instead of `Mono`, and if that make sense `Flow` instead of Flux.

simonbasle committed:
Add support for Kotlin suspending functions

Closes gh-28515

simonbasle committed:
Polish assertion messages and checkpoint message

simonbasle committed:
Document reactive/kotlin support in javadoc of the annotations

simonbasle committed:
Document reactive Scheduled support in the reference guide

simonbasle committed:
fix unclosed b tag in javadoc

sdeleuze reviewed (commented):
What if we would keep `ScheduledAnnotationReactiveSupport.isReactive(method)` and `processScheduledReactive(scheduled, method, bean)` API as they are and we just adapt the implementations to support suspending functions?

simonbasle committed:
Polish: use a single check for reactive and kotlin methods

The support's `isReactive` method checks Kotlin suspending functions
first then reactive (Publisher-returning) methods second. It asserts
the relevant runtime, all in a single call.

Similarly, turning the Method into a Publisher is done via a single
common helper method `getPublisherFor(Method, Object)`.

All imports of reactive classes and other reactive-specific logic is
still outsourced to ScheduledAnnotationReactiveSupport in order to avoid
any classpath issue in the bean postprocessor.

simonbasle commented:
> What if we would keep `ScheduledAnnotationReactiveSupport.isReactive(method)` and `processScheduledReactive(scheduled, method, bean)` API as they are and we just adapt the implementations to support suspending functions?

Done :) As stated in the latest commit, anything `Publisher` is still in `ScheduledAnnotationReactiveSupport`, but now there is only one check to perform (and the `getPublisherFor` method used by `ReactiveTask` automatically detect the correct case).

sdeleuze commented on spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationReactiveSupport.java:
```diff
@@ -0,0 +1,209 @@
+		if (!publisherPresent) {
+			return false;
+		}
+		if (!Publisher.class.isAssignableFrom(method.getReturnType())) {
```
May trigger a NPE if `Publisher` is not present in the classpath I think.

simonbasle replied:
the `if (!publisherPresent)` check above should prevent that part of the code to be reached and thus avoid a class-loading exception

sdeleuze commented on spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationReactiveSupport.java:
```diff
@@ -0,0 +1,209 @@
+	 * @throws IllegalStateException if the method is reactive but Reactor and/or the
+	 * Kotlin coroutines bridge are not present at runtime
+	 */
+	static boolean isReactive(Method method) {
```
Couldn't we use `ReactiveAdapterRegistry#getSharedInstance` to support various other non `Publisher` reactive types?

simonbasle replied:
I was under the impression that the canonical `ReactiveAdapterRegistry` isn't necessarily the `getSharedInstance` one but a bean one. When I started on this PR I initially tried to use a `ReactiveAdapterRegistry` from the context, but at the time the processing ran it wasn't available (which makes it difficult to detect arbitrary "reactive-convertible" types).

sdeleuze replied:
Given the fact that this feature needs to be implemented at `BeanPostProcessor` level and the fact that `ReactiveAdapterRegistry#getSharedInstance()` is pretty useful and flexible out of the box via its classpath detection mechanism, I think for this use case it is reasonable to have the limitation that it will not be customizable from the application context. Maybe worth documenting both at Javadoc and reference documentation level.

sdeleuze commented on spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationBeanPostProcessor.java:
```diff
@@ -386,14 +397,24 @@ public Object postProcessAfterInitialization(Object bean, String beanName) {
 		return bean;
 	}
 
```
Maybe copy the `processScheduled` original Javadoc here?

sdeleuze commented on spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationBeanPostProcessor.java:
```diff
@@ -386,14 +397,24 @@ public Object postProcessAfterInitialization(Object bean, String beanName) {
+	}
+
 	/**
 	 * Process the given {@code @Scheduled} method declaration on the given bean.
```
Maybe adapt the Javadoc to mention this is the sync variant?

simonbasle committed:
polish javadocs

simonbasle committed:
Use ReactiveAdapterRegistry to cover more types, detect wrong param count earlier

simonbasle committed:
Polish reference guide + annotation javadoc

simonbasle committed:
polish: remove FIXME

sdeleuze reviewed (approved):
Looks good, I just mentioned 2 possibles syles improvements in tests. Nice PR!

bclozel commented:
Thanks for tackling this - it's definitely the right time to address this as we'll work on observability in #29883.

simonbasle committed:
polish indentation

rstoyanchev reviewed (commented):
It's good to see this planned. 

My initial comment is to discuss the choice to use Reactor for scheduling. It means a more limited usage style, excluding cron expressions. Also, limits the usefulness of `ReactorAdapterRegistry` to a degree by requiring Reactor for the scheduling. I'm also wondering if `ReactiveTask` is going to cause classpath issues for sync methods but that's a more secondary, implementation concern.

Could the scheduling infrastructure used for sync also be used for reactive methods too, with some additional logic to subscribe to the Publisher every time a task is scheduled?

simonbasle committed:
Pivot to use the existing Runnable-based infrastructure

simonbasle committed:
polish and fix tests

simonbasle committed:
Track active subscriptions as Runnable in the processor

simonbasle commented:
I have attempted to take that approach which reuses a maximum of the existing infrastructure by using `Runnable` which subscribe to the Publishers. In a second step, I've attempted to reintroduce some tracking of active subscriptions. That way, `Publisher` subscriptions that don't terminate before the originating bean is destroyed will have a chance to be cancelled  (including infinite publishers).

simonbasle committed:
polish javadoc and fixmes

simonbasle commented:
I will now rebase on top of `main` to take the Antora doc changes into account, and will update the PR body to make it a relevant commit message for the current state of the PR.

simonbasle committed:
improve testability of the subscribing runnables

bclozel commented:
I'm currently working on the observability support of scheduled tasks in #29883. It looks like the current proposal for reactive methods support has a difference of behavior. Let's use two variants of scheduled methods, one blocking and the other reactive, both throwing exceptions during their processing:

```java
@Component
public class ScheduledComponent {

    private static final Logger logger = LoggerFactory.getLogger(ScheduledComponent.class);

    @Scheduled(cron = "0,10,20,30,40,50 * * * * *")
    public void blocking() {
        logger.info("Executing 'blocking' @Scheduled method");
        throw new IllegalStateException("Blocking method failed");
    }

    @Scheduled(cron = "2,12,22,32,42,52 * * * * *")
    public Mono<Void> reactive() {
        return Mono.error(() -> new IllegalStateException("Reactive method failed"))
                .doOnError(exc -> logger.info("Executing 'reactive' @Scheduled method"))
                .then();
    }
}
```

The blocking variant throws the exception from the generated runnable, and it's caught by the `org.springframework.util.ErrorHandler` infrastructure (this can be customized on `TaskScheduler` implementations):

```
2023-06-05T10:44:30.004+02:00  INFO 17840 --- [   scheduling-1] c.example.scheduling.ScheduledComponent  : Executing 'blocking' @Scheduled method
2023-06-05T10:44:30.004+02:00 ERROR 17840 --- [   scheduling-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task

java.lang.IllegalStateException: Blocking method failed
	at com.example.scheduling.ScheduledComponent.blocking(ScheduledComponent.java:17) ~[main/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:96) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
```

With the current proposal for reactive support, exceptions are never thrown and the error handling contract is not used:

```
2023-06-05T10:44:32.019+02:00  INFO 17840 --- [   scheduling-1] c.example.scheduling.ScheduledComponent  : Executing 'reactive' @Scheduled method
2023-06-05T10:44:32.023+02:00  WARN 17840 --- [   scheduling-1] s.s.a.ScheduledAnnotationReactiveSupport : Unexpected error occurred in scheduled reactive task

java.lang.IllegalStateException: Reactive method failed
	at com.example.scheduling.ScheduledComponent.lambda$reactive$0(ScheduledComponent.java:22) ~[main/:na]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ @Scheduled 'reactive()' in bean 'com.example.scheduling.ScheduledComponent'
Original Stack Trace:
		at com.example.scheduling.ScheduledComponent.lambda$reactive$0(ScheduledComponent.java:22) ~[main/:na]
		at reactor.core.publisher.MonoErrorSupplied.subscribe(MonoErrorSupplied.java:55) ~[reactor-core-3.5.6.jar:3.5.6]
		at reactor.core.publisher.Flux.subscribe(Flux.java:8671) ~[reactor-core-3.5.6.jar:3.5.6]
		at org.springframework.scheduling.annotation.ScheduledAnnotationReactiveSupport$SubscribingRunnable.run(ScheduledAnnotationReactiveSupport.java:194) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
		at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
		at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:96) ~[spring-context-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[na:na]
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
		at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
		at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]

2023-06-05T10:44:32.027+02:00 ERROR 17840 --- [   scheduling-1] reactor.core.publisher.Operators         : Operator called default onErrorDropped
```

I don't see this as a limitation, but really as a key difference between imperative and async in general. This behavior is already documented in this PR, but maybe we can mention that the `ErrorHandler` will not be invoked. From an observability perspective, I think we can probably still record failures with the observations.

bclozel commented:
Closed with 35052f2113274be
